pkg/controller/v1alpha1/llmisvc/workload_multi_node.go

```diff
@@ -0,0 +1,375 @@
+/*
+Copyright 2023 The KServe Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package llmisvc
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"knative.dev/pkg/kmeta"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	lwsapi "sigs.k8s.io/lws/api/leaderworkerset/v1"
+
+	"github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+)
+
+func (r *LLMInferenceServiceReconciler) reconcileMultiNodeWorkload(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) error {
+	log.FromContext(ctx).Info("Reconciling multi-node workload")
+
+	if err := r.reconcileMultiNodeMainServiceAccount(ctx, llmSvc); err != nil {
+		return fmt.Errorf("failed to reconcile multi-node service account: %w", err)
+	}
+	if err := r.reconcileMultiNodeMainWorkload(ctx, llmSvc); err != nil {
+		return fmt.Errorf("failed to reconcile multi-node main workload: %w", err)
+	}
+	if err := r.reconcileMultiNodePrefillWorkload(ctx, llmSvc); err != nil {
+		return fmt.Errorf("failed to reconcile multi-node prefill workload: %w", err)
+	}
+	return nil
+}
+
+func (r *LLMInferenceServiceReconciler) reconcileMultiNodeMainWorkload(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) error {
+	expected := r.expectedMainMultiNodeLWS(ctx, llmSvc)
+	if llmSvc.Spec.Worker == nil {
+		if err := Delete(ctx, r, llmSvc, expected); err != nil {
```

Comment:
Is this tested? Delete assume that llmSvc is already present

Reply:
Delete internally handles `NotFound` will not raise errors when the resource is not found.